### PR TITLE
Move ResourceManager to shared CoreLib.

### PR DESCRIPTION
**Pull request by filipnavara**
Posting for early feedback before I proceed to adapt the CoreRT part. Contributes to #21932

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs [hunk 1]
```diff
@@ -566,9 +551,25 @@ protected static Version GetSatelliteContractVersion(Assembly a)
...
+            {
+                // Note we are prone to hitting infinite loops if mscorlib's
+                // SatelliteContractVersionAttribute contains bogus values.
+                // If this assert fires, please fix the build process for the
+                // BCL directory.
+                Debug.Assert(a == typeof(object).GetTypeInfo().Assembly, System.CoreLib.Name + "'s SatelliteContractVersionAttribute is a malformed version string!");
+                throw new ArgumentException(SR.Format(SR.Arg_InvalidSatelliteContract_Asm_Ver, a.ToString(), v));
+            }
+
+            return version;
```
Comment (filipnavara) [inline src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs]: The code for `SatelliteContractVersionAttribute` was not present in CoreCLR. There is issue filed for it - #17816 - and I am not sure whether there was some specific reason to omit it.

File: src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs [hunk 2]
```diff
@@ -122,11 +113,6 @@ internal class CultureNameResourceSetPair
         }
 
         protected string BaseNameField;
-        // Sets is a many-to-one table of CultureInfos mapped to ResourceSets.
-        // Don't synchronize ResourceSets - too fine-grained a lock to be effective
-        [Obsolete("call InternalGetResourceSet instead")]
-        internal Hashtable ResourceSets;
```
Comment (filipnavara) [inline src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs]: This was not used anywhere and seems like a leftover from NetFX. It doesn't exist in CoreRT.

File: src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs [hunk 3]
```diff
@@ -250,6 +233,8 @@ public ResourceManager(string baseName, Assembly assembly, Type usingResourceSet
                 throw new ArgumentException(SR.Arg_ResMgrNotResSet, nameof(usingResourceSet));
             _userResourceSet = usingResourceSet;
 
+            SetAppXConfiguration();
+
```
Comment (filipnavara) [inline src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs]: This SetAppXConfiguration call was missing in CoreCLR, but present in CoreRT. It may have been intentional, but in that case it should still be here with `if (_userResourceSet == null)`. Certainly the calls to SetAppXConfiguration are inconsistent.

File: src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems [hunk 4]
```diff
@@ -1051,6 +1055,10 @@
     <Compile Include="$(MSBuildThisFileDirectory)System\Threading\Semaphore.Windows.cs" />
     <Compile Include="$(MSBuildThisFileDirectory)System\Threading\EventWaitHandle.Windows.cs" />
   </ItemGroup>
+  <ItemGroup Condition="$(TargetsWindows) and ('$(FeatureAppX)' == 'true' or '$(EnableWinRT)' == 'true')">
+    <Compile Include="$(MSBuildThisFileDirectory)Internal\Resources\PRIExceptionInfo.cs" />
+    <Compile Include="$(MSBuildThisFileDirectory)Internal\Resources\WindowsRuntimeResourceManagerBase.cs" />
+  </ItemGroup>
```
Comment (filipnavara) [inline src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems]: Moving these two files may or may not be ok. The CoreRT versions are in different namespace and the `Initialize` method has one less argument. Would it be possible to move it and share with ProjectN?

Comment (filipnavara) [reply src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems]: Funny story, there's this comment in CoreRT:

```
    // It is important to have WindowsRuntimeResourceManagerBase as regular class with virtual methods and default implementations. 
    // Defining WindowsRuntimeResourceManagerBase as abstract class or interface will cause issues when adding more methods to it 
    // because it'll create dependency between mscorlib and System.Runtime.WindowsRuntime which will require always shipping both DLLs together. 
    // Also using interface or abstract class will not play nice with FriendAccessAllowed.
```

Ironically it seems that these default implementations avoid an error when linking to CoreFX's current System.Runtime.WindowsRuntime. The `Initialize` implementation with 2 arguments has default implementation in CoreRT that does nothing and it is never overridden.

Am I missing something?

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems]: >  which will require always shipping both DLLs together

This part of the comment was added in an era where we believe that we can ship the framework as 100+ little individually versioned packages. It turned out to be not workable. There .dlls always ship together. You can delete this comment.

> Initialize implementation with 2 arguments has default implementation in CoreRT that does nothing and it is never overridden.

It is overriden in assemblies specific to the .NET Native for UWP that have not been open sourced.

Comment (filipnavara) [reply src/System.Private.CoreLib/shared/System.Private.CoreLib.Shared.projitems]: > It is overriden in assemblies specific to the .NET Native for UWP that have not been open sourced.

Let me rephrase that questions then.

Is it possible to update the non-open source version to have the same signatures as the version in CoreFX and move them to same namespace? (ie. adding one parameter to `Initialize` method and change namespace from `System.Resources` to `Internal.Resources`)

If yes, is it desirable to share the "interface" files?

If the answer to either of those two questions is "no" then I will just drop this part and move the original definitions in CoreRT to a separate non-shared file and add `Initialize` method there with forward to the 2-argument version.

File: src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs [hunk 5]
```diff
@@ -630,27 +630,37 @@ private string GetStringFromPRI(string stringName, string startingCulture, strin
...
+            Type WinRTResourceManagerType = hiddenScopeAssembly.GetType("System.Resources.WindowsRuntimeResourceManager", true);
+            return (WindowsRuntimeResourceManagerBase)Activator.CreateInstance(WinRTResourceManagerType, true);
+        }
+#endif
 
         private bool _bUsingModernResourceManagement; // Written only by SetAppXConfiguration
 
-#if FEATURE_APPX
+#if FEATURE_APPX || ENABLE_WINRT
         private WindowsRuntimeResourceManagerBase _WinRTResourceManager; // Written only by SetAppXConfiguration
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs]: >  Uap-specific code into a separate file 

Maybe some of the code from `CultureInfo.Windows.cs` can be in `ResourceManager.Uap.cs` too to simplify the ifdefs there.

File: src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs [hunk 6]
```diff
@@ -566,9 +551,25 @@ protected static Version GetSatelliteContractVersion(Assembly a)
...
             }
 
-            // Return null. The calling code will use the assembly version instead to avoid potential type
-            // and library loads caused by CA lookup. NetCF uses the assembly version always.
-            return null;
+            string v = a.GetCustomAttribute<SatelliteContractVersionAttribute>()?.Version;
+            if (v == null)
+            {
+                // Return null. The calling code will use the assembly version instead to avoid potential type
+                // and library loads caused by CA lookup. NetCF uses the assembly version always.
```
[the remark this answers is not in the document]
Comment (filipnavara) [reply src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs]: It's been a long time since I used NetCF :-) I will remove it and check the other comments for relevancy too.

File: src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (filipnavara) [reply src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs]: I did the basic split, but I would rather avoid doing some bigger refactoring as part of this PR.

File: src/System.Private.CoreLib/shared/System/Resources/ResourceManager.NonUap.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+namespace System.Resources
+{
+    public partial class ResourceManager
+    {
+        private bool UseUapResourceManagement { get => false; }
+
+        private string GetStringFromPRI(string stringName, CultureInfo culture, string neutralResourcesCulture) { return null; }
+
+        private void SetAppXConfiguration() { }
+    }
```
Comment (filipnavara) [inline src/System.Private.CoreLib/shared/System/Resources/ResourceManager.NonUap.cs]: I can get rid of this file and put two `#if`s in ResourceManager.cs instead if that is preferred.

File: src/System.Private.CoreLib/shared/System/Resources/ManifestBasedResourceGroveler.cs [hunk 8]
```diff
@@ -384,11 +384,13 @@ private Assembly GetSatelliteAssembly(CultureInfo lookForCulture)
                 // Ignore cases where the loader gets an access
                 // denied back from the OS.  This showed up for
                 // href-run exe's at one point.  
+#if CORECLR
                 int hr = fle.HResult;
                 if (hr != Win32Marshal.MakeHRFromErrorCode(Interop.Errors.ERROR_ACCESS_DENIED))
                 {
                     Debug.Fail("[This assert catches satellite assembly build/deployment problems - report this message to your build lab & loc engineer]" + Environment.NewLine + "GetSatelliteAssembly failed for culture " + lookForCulture.Name + " and version " + (_mediator.SatelliteContractVersion == null ? _mediator.MainAssembly.GetName().Version.ToString() : _mediator.SatelliteContractVersion.ToString()) + " of assembly " + _mediator.MainAssembly.GetName().Name + " with error code 0x" + hr.ToString("X", CultureInfo.InvariantCulture) + Environment.NewLine + "Exception: " + fle);
                 }
+#endif
```
Comment (filipnavara) [inline src/System.Private.CoreLib/shared/System/Resources/ManifestBasedResourceGroveler.cs]: Not sure this is worth keeping at all (or testing for ERROR_ACCESS_DENIED specifically).

Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Resources/ManifestBasedResourceGroveler.cs]: I think it would be fine to delete it.

File: src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/shared/System/Resources/ResourceManager.cs]: ok, we can fix the names in a separate PR

**Comment by marek-safar** [comment]
@jkotas could you land this PR?
